quickwit/quickwit-ingest-api/src/doc_batch.rs

```diff
@@ -0,0 +1,420 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use bytes::buf::Writer;
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use serde::Serialize;
+
+use crate::metrics::INGEST_METRICS;
+use crate::DocBatch;
+
+#[derive(Debug)]
+/// Represents a command that can be stored in doc batch.
+pub enum DocCommand<T>
+where T: Buf
+{
+    Ingest { payload: T },
+    Commit,
+    // ... more to come?
+}
+
+/// We can use this byte to track both commands and their version changes
+/// If serialization protocol changes, we can just use the next number
+const INGEST_CODE_V1: u8 = 0;
+const COMMIT_CODE_V1: u8 = 1;
```

Comment:
this should probably be a `repr(u8)` enum

Reply:
I might be missing something, but it didn't look like it would be worth it. If DocCommand was a plain enum that we could have easily casted it to u8, but since it is using generics and payload, I couldn't find any way of getting this code except through https://doc.rust-lang.org/std/mem/fn.discriminant.html and I wasn't sure if introducing unsafe code was justified here.

Reply:
I was thinking of a 2nd enum mirroring `DocCommand`, instead of a bunch of free-floating constants

Reply:
Great idea!